### PR TITLE
Roe 2355 improve model unit testing

### PR DESCRIPTION
**Pull request by mwestacott**
### JIRA link

https://companieshouse.atlassian.net/browse/ROE-2355

### Change description

_The current problem:_

If a DAO/DTO model field datatype changes (e.g. from `String` to `EntityNameDTO`) that causes a breaking change when reading older JSON model data from Mongo (requiring a new data transformer to be written), this is not picked up by the current `DtoModelChangeTest` unit-test unless the data field is present in the test JSON file. If not present and not detected by developers, it can lead to a later run-time failure, possibly in the LIVE environment.

_The proposed solution:_

This PR implements unit-test changes to ensure that (nearly) all fields present in the DTO/DAO model are present in the current (most recent) JSON test file, thereby ensuring that the model conversions will be exercised and tested by the `DtoModelChangeTest` unit-test. In addition, it updates the current JSON test file with missing data fields, so that the new test does not fail.  

### Work checklist

- [ ] Tests added where applicable

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/model/DtoModelChangeTest.java [hunk 1]
```diff
@@ -15,37 +16,75 @@
...
+     * If the DTO model changes and it is a 'breaking change', the current filename must be added to the
+     * <code>HISTORICAL_JSON_MODEL_FILENAMES</code> list and a new test JSON file will need to be created and then
+     * referenced here.
+     */
+    private static final String CURRENT_JSON_MODEL_FILENAME = "overseas_entity_v_3_1.json";
+
+    /**
+     * Allows the list to be used as input to a JUnit parameterised test (see below).
+     */
+    private static List<String> getHistoricalJsonModelFilenames() {
```
Comment (mattch1) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/model/DtoModelChangeTest.java:84]: Could just declare a static array list and then below have:

static { 
    arrayList.add("overseas_entity_v_1_0.json")
    ...
}
etc.


}

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/model/DtoModelChangeTest.java [hunk 2]
```diff
@@ -15,37 +16,75 @@
...
- * model structure should be added to the test resources folder and a new test added to successfully read and process
- * the new, up-to-date DAO/DTO structure.
+ * If a non-breaking change to the model has been made - e.g. a new data field has simply been added to a DTO and DAO -
+ * the 'current' test JSON file needs updating to avoid a null check failure (this essentially future-proofs the test,
+ * which will potentially fail if the datatype of the newly added field ever changes).
  *
- * Note the use of Spring profiles, to ensure that the (default) <code>MongoConfig</code> configuration is not loaded
- * during this test (as the specific in-memory Mongo DB config needs to be picked up).
+ * If a breaking change to the model has been made - e.g. a data field has changed its type from String to a custom POJO -
+ * after a new 'transformer' has been implemented, a file with the new JSON model structure should be added to the test
```
Comment (markpit) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/model/DtoModelChangeTest.java:55]: i would also mention that a new SchemaVersion would need to be added alongside the transformer

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/model/DtoModelChangeTest.java [hunk 3]
```diff
@@ -15,37 +16,75 @@
...
+ * resources folder and the <code>OLD_JSON_MODEL_FILENAMES</code> and <code>CURRENT_JSON_MODEL_FILENAME</code> constants
+ * updated, in order that the new, up-to-date DAO/DTO structure is fully tested.
+ *
+ * Finally, note the use of Spring profiles, to ensure that the (default) <code>MongoConfig</code> configuration is not
+ * loaded during this test (as the specific in-memory Mongo DB config needs to be picked up).
  */
 @AutoConfigureDataMongo
 @SpringBootTest
 @ActiveProfiles("test")
```
Comment (markpit) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/model/DtoModelChangeTest.java:64]: i know it's not new but this test class seems to be the only one with @ActiveProfiles("test"), perhaps remove it so it always runs with the other tests?

Comment (mwestacott) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/model/DtoModelChangeTest.java:64]: It's (still) needed - see comment on line 59. If not present, test fails to initialise and run.

File: src/test/resources/overseas_entity_v_3_1.json [hunk 4]
```diff
@@ -504,6 +612,100 @@
         "resigned_on" : "2009-08-01",
         "ch_reference" : "MOcorp1"
       }
+    ],
```
Comment (jkayyala) [inline src/test/resources/overseas_entity_v_3_1.json:615]: The model does not contain update object @AndreKramer7654 will that needed to be added in another ticket

File: src/test/resources/overseas_entity_v_3_1.json [hunk 5]
```diff
@@ -22,7 +23,10 @@
       "line_2" : "Line 2",
       "town" : "Town",
       "county" : "County",
+      "locality" : "Locality",
       "country" : "Aland Islands",
+      "po_box" : "PO Box",
```
[the remark this answers is not in the document]
Comment (mwestacott) [reply src/test/resources/overseas_entity_v_3_1.json:28]: Yes, definitely worth pointing out in this PR as we weren't aware of this. Now we are and I can get a tech debt ticket raised (somewhere). Having spoken to @SteveHicksCH, it seems that these were put in for trusts (originally present in the spreadsheet) but then deemed not necessary when the web screens were implemented.

In the meantime, adding data in the JSON for these fields for this test still needs to be done as the fields are there in the DAO/DTO model.

File: src/test/resources/overseas_entity_v_3_1.json (hunk 4 above)
Comment (mwestacott) [reply src/test/resources/overseas_entity_v_3_1.json:615]: I'm confused. Isn't that structure present from line 327 of this file?(I think @markpit added this fairly recently, to be compliant). It's there as a field in the `OverseasEntitySubmissionDto` too.

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/model/DtoModelChangeTest.java (hunk 2 above)
Comment (mwestacott) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/model/DtoModelChangeTest.java]: Changes pushed with updated comments.

File: src/test/resources/overseas_entity_v_3_1.json (hunk 4 above)
Comment (jkayyala) [reply src/test/resources/overseas_entity_v_3_1.json:615]: Seems I missed that ...

**Review by jkayyala — APPROVED** [review]

**Review by markpit — APPROVED** [review]

**Pull request merged**